tests/system/libraries/systemTestSpy.py

```diff
@@ -0,0 +1,207 @@
+		return f
+
+	def _getJoinedBaseStringsFromCommands(self, speechCommandArray):
+		wsChars = whitespaceMinusSlashN
+		baseStrings = [c.strip(wsChars) for c in speechCommandArray if isinstance(c, basestring)]
+		return ''.join(baseStrings).strip()
+
+	# Public methods
+	def checkIfSpeechOccurredAndReset(self):
+		# don't let _speechOccurred get updated and overwritten with False
+		with threading.Lock():
+			speechOccurred = self._speechOccurred
+			self._speechOccurred = False
+		return speechOccurred
+
+	def getSpeechAtIndex(self, speechIndex):
+		with threading.Lock():
+			return self._getJoinedBaseStringsFromCommands(self._nvdaSpeech[speechIndex])
+
+	def getSpeechSinceIndex(self, speechIndex):
+		with threading.Lock():
+			speechCommands = self._flattenCommandsSeparatingWithNewline(
+				self._nvdaSpeech[speechIndex:]
+			)
+			joined = self._getJoinedBaseStringsFromCommands(speechCommands)
+			return joined
+
+	def getIndexOfLastSpeech(self):
+		with threading.Lock():
+			return len(self._nvdaSpeech) - 1
+
+	def getIndexOfSpeech(self, speech, startFromIndex=0):
+		with threading.Lock():
+			for index, commands in enumerate(self._nvdaSpeech[startFromIndex:]):
+				index = index + startFromIndex
+				baseStrings = [c.strip() for c in commands if isinstance(c, basestring)]
+				if any(speech in x for x in baseStrings):
+					return index
+			return -1
+
+	def hasSpeechFinished(self):
+		return self.SPEECH_HAS_FINISHED_SECONDS < _timer() - self.lastSpeechTime
+
+	def dumpSpeechToLog(self):
+		log.debug("All speech:\n{}".format(repr(self._nvdaSpeech)))
+
+class NvdaSpyLib(object):
+	_spy = None  # type: SystemTestSpy
+
+	def __init__(self, systemTestSpy):
+		self._spy = systemTestSpy
+		self._allSpeechStartIndex = self._spy.getIndexOfLastSpeech()
+		self._maxKeywordDuration=30
+
+	def _minTimeout(self, timeout):
+		"""Helper to get the minimum value, the timeout passed in, or self._maxKeywordDuration"""
+		return min(timeout, self._maxKeywordDuration)
+
+	# Start of Robot library API
+
+	def set_max_keyword_duration(self, maxSeconds):
+		"""This should only be called after importing the library, and should match the 'timeout' value given to the
+		robot.libraries.Remote instance"""
+		self._maxKeywordDuration = maxSeconds-1
+
+	def wait_for_NVDA_startup_to_complete(self):
+		_blockUntilConditionMet(
+			getValue=lambda: self._spy.isNvdaStartupComplete,
+			giveUpAfterSeconds=self._minTimeout(10),
+			errorMessage="Unable to connect to nvdaSpy",
+		)
+		if self._spy.isNvdaStartupComplete:
+			self.reset_all_speech_index()
+
+	def get_last_speech(self):
+		return self._spy.getSpeechAtIndex(-1)
+
+	def get_all_speech(self):
+		return self._spy.getSpeechSinceIndex(self._allSpeechStartIndex)
+
+	def get_speech_from_index_until_now(self, speechIndex):
+		return self._spy.getSpeechSinceIndex(speechIndex)
+
+	def reset_all_speech_index(self):
+		self._allSpeechStartIndex = self._spy.getIndexOfLastSpeech()

```

Comment:
The name suggests this should reset the speech. Does this actually do anything but return the last index?

Reply:
Hmm, perhaps it's because its not clear that "all speech" is a grouping. This resets the "all speech" index, meaning that when you next call `get_all_speech` it will return any new speech added after the index was reset. It's important that we don't actually clear the internal buffer, since tests may need to hold on to speech indexes and be able to refer to them later, without them being invalidated.